### ISSUE TITLE
[D8] Allow forms to set custom validation error messages on required fields

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5348]
This is the respective issue for https://www.drupal.org/project/drupal/issues/742344, a blocker for the #5346 backport, and related to #1040.

> ## Problem/Motivation
> Accessibility guidelines recommend specific, meaningful error messages when required form elements are empty. However, FAPI currently does not support custom error messages for required elements.
> 
> ## Proposed resolution
> Support a `#required_error` property for form elements. If set, this message will be used when a required form element is empty, instead of the default "Field x is required."
> Additionally set a `#required_but_empty` flag automatically during validation when the required field is empty. This is an internal property and is not set manually. It allows `#element_validate` handlers to set a custom required error message, but without having to re-implement the complex logic to figure out whether the field value is empty.

> ## Remaining tasks
> Commit feature and document these properties. The feature is a small API improvement that is applicable to both D8 and D7.
> 
> ## User interface changes
> None.

> ## API changes
> Minor, backwards-compatible addition only. No existing code will be affected. Addition of two FAPI element properties:
> 
> - `#required_error`
> - `#required_but_empty`

---

Advocate for this issue: @klonos

**Comment by klonos** [comment]
> Minor, backwards-compatible addition only. No existing code will be affected. Addition of two FAPI element properties...

Can this be regarded as a task instead of a feature then? I'd love to have this (and also #5346) in place before starting work on #1040 for the 1.21.0 release.

**Comment by klonos** [comment]
To get this moving, in order to progress #1040, I've set the milestone label for the next minor release. Can I please get someone to add the actual milestone? Thanks.

PR ready for review/testing: https://github.com/backdrop/backdrop/pull/3844
Draft/unpublished change record here: https://docs.backdropcms.org/change-records/new-form-api-properties-required_error-and-required_but_empty

**Comment by bugfolder** [comment]
Tested the PR with this form:
```
/**
 * Return a test form for custom errors.
 */
function backdrop_test_custom_error_form($form, &$form_state) {
  $form['backdrop_test_custom_error'] = array(
    '#type' => 'textfield',
    '#title' => t('Custom error test'),
    '#required' => TRUE,
    '#required_error' => t('You must fill in a value for the field.'),
//    '#element_validate' => array('backdrop_test_custom_error_form_validate'),
  );

  $form['submit'] = array(
    '#type' => 'submit',
    '#value' => t('Submit'),
    '#suffix' => l(t('Cancel'), 'backdrop-test-custom-error'),
  );
  return $form;
}

/**
 * Validation for custom error form testing.
 */
function backdrop_test_custom_error_form_validate($element, &$form_state) {
  if (isset($element['#required_but_empty'])) {
    form_error($element, t('Someone must fill in a value for the field.'));
  }
}
```
Worked as expected, both as above and with `#element_validate` uncommented. 

Reviewed the code: LGTM.

**Comment by klonos** [comment]
> Worked as expected...

Thanks @bugfolder 🙏🏼  ...based on your comment above, I've also se the "WFM" label.

**Comment by argiepiano** [comment]
Hey @klonos . I haven't had a chance to test this, but at quick glance, this PR may break a test that I added in PR  https://github.com/backdrop/backdrop/pull/3837 that fixed the issue with the field label not being displayed in the validation error for multi-value required fields. That test is in `field.test` in method  `testFieldFormUnlimitedRequired()`.

I'm not sure about this, but there may be other tests in that file that check for validation error specific wording that may also be affected. Sorry, I don't have time until tonight to check, but you may have caught those other ones already?

**Comment by klonos** [comment]
> ...this PR may break a test that I added in PR  https://github.com/backdrop/backdrop/pull/3837

That PR was merged a couple of hours ago, so let me rebase and test this again...

**Comment by klonos** [comment]
> (Although on second thought this may not be an issue as those tests do not mess with `#required_error` and therefore the default validation error will be shown)

Well, most likely yes, but lets make sure. ...I've rebased and pushed. Tests are currently running, so let's see.

**Comment by klonos** [comment]
...OK, this was a clean rebase (no conflicts to resolve - no changes pushed), and all tests are still green. So we're good 👍🏼

**Comment by klonos** [comment]
Huh! ...interesting find while working on #5381:

`search_box_form_submit()`:
```php
...

  // Check to see if the form was submitted empty.
  // If it is empty, display an error message.
  // (This method is used instead of setting #required to TRUE for this field
  // because that results in a confusing error message. It would say a plain
  // "field is required" because the search keywords field has no title.
  // The error message would also complain about a missing #title field.)
  if ($form_state['values']['search_block_form'] == '') {
    form_set_error('keys', t('Please enter some keywords.'));
  }

...
```

**Comment by klonos** [comment]
...yes, this actually seems like the perfect use case in core for the feature we're introducing here!

1. The inline comment has the following 2 statements, none of which are true:
   - `It would say a plain "field is required" because the search keywords field has no title.`
   - `The error message would also complain about a missing #title field.`

   I believe that this comment must have been written during a time where the label of the search field was completely removed from the form. We now have `'#title' => t('Search')` and `'#title_display' => 'invisible'` in place.

   So, if you add `'#required' => TRUE` to the search field now, the error says "Search field is required.":
   ![image](https://user-images.githubusercontent.com/2423362/144709503-fb72cef6-99fc-4692-bcff-81059689eba1.png)

   It even says "Search field is required." when the block has been configured to have its label hidden:
   ![image](https://user-images.githubusercontent.com/2423362/144709055-14652b64-4405-4a3d-a7f2-685428acedbf.png)

So I have removed this legacy pseudo-required workaround from `search_box_form_submit()`, have added `'#required' => TRUE` to the search field, and have implemented the newly-introduced `#required_error` property. Here's what it looks like now (the custom "Please enter some keywords..." message is shown, same as previously, but without the silly workaround):

![image](https://user-images.githubusercontent.com/2423362/144711047-5b53ebef-fe2f-4984-becc-a03375b3bbdf.png)

...as an added UX bonus, the form validation now works consistently, same as other forms with required fields:
- if you leave the field empty, you see the validation error in the same page where you clicked the "Search" button from (as opposed to being redirected to `search/node` when the form validation happens).
- if you enter some search keywords, the redirection to `search/node` still works as previously.

**Comment by bugfolder** [comment]
Re-tested with example above https://github.com/backdrop/backdrop-issues/issues/5348#issuecomment-985089399. 
Still worked:

* `#required_error`(but not `#element validate`) generated the appropriate error message. 
* `#element_validate` (but not `#required_error`) with validator checking `#required_but_empty` generated the appropriate error message.
* Having both `#required_error` and `#element_validate` generated just the `#required_error` error message.
* Having neither generated the generic default error message.

Also tested the Search page `/search`; empty search resulted in "Please enter some keywords." Note, not "Please enter some keywords in the search field," as above. This error message is generated in function `search_form_submit()`. Should we use the same language in both `search_box()` and `search_form_submit()`?

**Comment by klonos** [comment]
> Should we use the same language in both `search_box()` and `search_form_submit()`?

We sure should 👍🏼  ...please check the latest commit: [backdrop/backdrop@`6130868` (#3844)](https://github.com/backdrop/backdrop/pull/3844/commits/613086822ffdb46c27b10bc730d809a22a6aecba)

Once this feature gets merged into core, we should do some cleanup in all our `*_validate()` and `*_submit()` functions, to move any custom `form_set_error()` logic from those into the new `'#required_error'` property for the respective required fields. But that needs to be a follow-up (or we can slowly be converting those as we go).

In general, although I have created #5381 and this issue here as "prep work" for #1040, I think that these features will provide the means to do some serious codebase clean-up + provide consistency across the way we handle form validation. Less work for developers + consistency -> better DX.

**Comment by bugfolder** [comment]
Tested. WFM.

**Comment by klonos** [comment]
I actually found something related to the search form that might be broken. Details coming soon...

**Comment by docwilmot** [comment]
@klonos is there a reason why you chose to port this D7 patch? There seem to be consensus that it broke field validation, and an alternative patch was committed after this one AFAICT.

**Comment by docwilmot** [comment]
I'm also not seeing why this is a blocker for https://github.com/backdrop/backdrop-issues/issues/5346 or https://github.com/backdrop/backdrop-issues/issues/1040

**Comment by quicksketch** [comment]
This issue isn't necessarily a blocker for #5346, but it's required if we want to keep any custom required messages and use HTML5 `required` attributes together. It may also be helpful if we were to ever add a server-side conditional system to replace `#states`. By separating out the required message into it's own property, we could conditionally display the required message without executing the entire form `_validate()` handler.

> I actually found something related to the search form that might be broken. Details coming soon...

Any update on this?

**Comment by quicksketch** [comment]
Bumping this to 1.22.0. Thanks @klonos for your work on this and I hope we can get it completed soon.

**Comment by docwilmot** [comment]
We're not doing this? Its a core milestone *and* a contrib candidate?

**Comment by jenlampton** [comment]
The contrib candidate label indicates that this feature is a good candidate for being a contrib module instead of a core change. Since I added it back in march, I'll do another read through and add a comment as to why I thought that applied here.

**Comment by quicksketch** [comment]
Code here looks pretty good. I reviewed and the implementation seems reasonable. I left a question on the PR: https://github.com/backdrop/backdrop/pull/3844#pullrequestreview-951266130

I also merged in 1.x to test against the latest code, including bumping up to PHP 5.6 (minimum) and 8.1 (max).

The only test that is failing is related to the advanced search form, where the new `#required_error` is used. This is because the advanced node form is tested in a situation where the main search terms can be left blank, and the advanced search terms are used instead, such as in the following screenshot:

![image](https://user-images.githubusercontent.com/100206/165008656-b00ec1d3-d7cb-4123-a921-9d2753f54d5a.png)

In this situation, the search field is not actually required because it gets populated by other fields. It's totally confusing and not the way I think we'd handle this if we reworked this form, but I think we should simply revert the search changes to avoid this issue.

@klonos: suggestion for naming: How would you feel about `#required_message` instead of `#required_error`? I feel like "message" more clearly indicates a string value, where "error" might imply a boolean TRUE/FALSE value.

**Comment by quicksketch** [comment]
Naming convention-wise, we can try to line up with the JavaScript Constraint Validation API: https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#the_constraint_validation_api

It uses `validationMessage` (for any failed validation, depending on the situation) and `valueMissing` (for a required-but-empty value).

So I'd be in favor of the following property names:

* `#required_message` (instead of the current `#required_error`)
* `#required_value_missing` (instead of the current `#required_but_empty`)

**Comment by quicksketch** [comment]
I filed an updated PR at https://github.com/backdrop/backdrop/pull/4054 that makes the above property name changes and reverts _some_ of the search changes. The search block can use the new `#required_message` property but the main search form cannot, because the main terms field is optional and may be populated by the advanced search fields instead. I added a code comment explaining why the field is not marked required.

**Comment by docwilmot** [comment]
I mentioned in https://github.com/backdrop/backdrop-issues/issues/5348#issuecomment-998442599 above that the patch this was based on was superceded in D7 by a later patch. Particularly, [this comment](https://www.drupal.org/project/drupal/issues/742344#comment-6690418) about the initial patch made sense. In short, if we have a required field (must not be empty) with *also* a custom validation, if the field is submitted empty we'd only want the "field is empty" message to trigger, because the custom validation is then not useful. If the field is *not* empty, only then do we want our custom validation to run.

If I'm not mistaken this PR has this weakness still.

**Comment by quicksketch** [comment]
Thanks @docwilmot, good catch and I admit I did not understand why the PR had moved the order of validation around. We cannot fundamentally change the flow of Form API. If `#element_validate` only fired after required values were provided previously, we need to maintain that expectation.

I updated my PR, which is now less changes than before. A side-effect of this is that now I don't see any reason to set a `#required_value_missing` property, since there are no other validate callbacks that could make use it. If a required field is not filled out, neither `#element_validate` nor overall form `#validate` callbacks are called.

I also added one more test to ensure the relationship between `#element_validate` and `#required` is maintained as it is today.

**Comment by quicksketch** [comment]
I'm removing contrib candidate because I don't think this feature would be useful in contrib. This is a convenience ability that improves Form API, and it wouldn't make sense to have modules depend on it as a stand-alone feature. It would only make sense as part of a larger suite of functionality like CTools used to provide.

**Comment by indigoxela** [comment]
Works for me, is how I tested...

A custom mini-module just for testing:
```
function foobar_form_node_form_alter(&$form, &$form_state, $form_id) {
  $form['favpet'] = array(
    '#type' => 'textfield',
    '#title' => t('Favorite pet'),
    '#default_value' => '',
    '#weight' => -5,
    '#required_message' => t('Please enter your favorite pet (the species).'),
    '#required' => TRUE,
  );
}
```
And if I leave the field empty when adding a node, I get:
![custom-empty-message](https://user-images.githubusercontent.com/842066/166142482-979e18e8-0c21-4d00-9fcd-780f8ea182dc.png)
:+1: 

Without that PR (or the new property set), I'd simply get the usual "Favorite pet field is required."

**Comment by docwilmot** [comment]
Took me a while to understand why we added a custom `#form_test_required_message`, but I realise this just holds the expected message for that test, right? As a followup, perhaps rename that to `#form_test_expected_message` and a docblock to say this is not a standard API key, but just added for that purpose.

Otherwise, LGTM.

**Comment by quicksketch** [comment]
Thanks @docwilmot and @indigoxela! Your testing and feedback is much appreciated.

@docwilmot I went ahead and changed that property name as you suggested and added code comments.

Further, I reverted more of the original changes to Search module because they caused a regression in intentional behavior. If the user clicks "Search" but does not enter any terms, the intentional behavior seems to have been to take the user to the /search page before showing the error message. This makes sense, and so I left the behavior in place but updated the code comments to explain why `#required` is not used.

**Comment by docwilmot** [comment]
Only one small language nitpick.

**Comment by quicksketch** [comment]
We didn't get this finalized in time for 1.22.0, so I'm bumping this to the next release.

**Comment by bugfolder** [comment]
Code reviewed. Aside from a few tiny wording/case suggestions in comments, all looks good.

**Comment by klonos** [comment]
> So I'd be in favor of the following property names:
> 
> * `#required_message` (instead of the current `#required_error`)
> * `#required_value_missing` (instead of the current `#required_but_empty`)

Sorry for not chiming in for so long here ...I chose to go with `#required_error` and `#required_but_empty` in order to keep some party with the equivalent changes that are happening in Drupal-land. My thinking was that if these are somehow used in contrib and/or any documentation/tutorials or code snippets, then perhaps having the same naming conventions as Drupal would make it that less work is required for cross-/back-ports. Happy for us to go wither way, and the proposed new names based on what these things are called in the JavaScript Constraint Validation API do sound better IMO 👍🏼 

>> I actually found something related to the search form that might be broken. Details coming soon...
>
> Any update on this?

...

> ...the search field is not actually required because it gets populated by other fields. It's totally confusing and not the way I think we'd handle this if we reworked this form...

If memory serves well (been a few months since), that's the search-related issue I mentioned I came across back in December last year. Totally confusing as you said @quicksketch, and at the time I didn't know how to proceed with it.

> The search block can use the new `#required_message` property but the main search form cannot, because the main terms field is optional and may be populated by the advanced search fields instead. I added a code comment explaining why the field is not marked required.

Yes 🤔 ...I think that we should file a follow up issue to eventually fix things properly with the search form, but I suspect that we'd need to fix things like #2732 and perhaps other issues related to how required fields are (not) handled by `#states`.

I'll need to have a better look at the code changes, but please do not hold this on account of me.

**Comment by indigoxela** [comment]
@bugfolder you removed the "needs code review" label, but didn't add "rtbc" - why that?

**Comment by bugfolder** [comment]
Overlooked it.

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/4054 into 1.x for 1.23.0! Thanks folks!

https://github.com/backdrop/backdrop/commit/7e760e91f32e65d4c8d11471232fd4ffd2fde906 by @klonos, @bugfolder, @quicksketch, @docwilmot, and @indigoxela.
